Financial multilingual embed notebook

gonzalocohere opened the pull request:
Added notebook with walkthrough of multilingual embeddings for searching financial data using Bedrock. It also has a final Rerank component using SageMaker (to be available in Bedrock).

gonzalocohere committed:
added bedrock example notebook

Example leveraging multilingual embedding model for financial data using Bedrock and then reranking with SageMaker

gonzalocohere committed:
Rename Financial Multlingual Embeddings.ipynb to Financial Multilingual Embeddings.ipynb

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/cohere-ai/cohere-aws?pullRequest=88) <br/>All committers have signed the CLA.

leilacc commented on notebooks/bedrock/Financial Multilingual Embeddings.ipynb:
```diff
@@ -0,0 +1,861 @@
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Requirement already satisfied: cohere-aws in /home/ec2-user/anaconda3/envs/python3/lib/python3.10/site-packages (0.8.10)\n",
```
Can you clear the output? & ditto for the output in other sections

gonzalocohere replied:
Fixed in 91f7d1cb406b3cadda57438b1e8c1ecb8d7f89eb

leilacc commented on notebooks/bedrock/Financial Multilingual Embeddings.ipynb:
```diff
@@ -0,0 +1,861 @@
+   "id": "494c504f-0886-4b29-8716-0a47e6450647",
+   "metadata": {},
+   "source": [
+    "### Step 1: Install Packages and Import Modules|"
```
trailing |

gonzalocohere replied:
Fixed in 91f7d1cb406b3cadda57438b1e8c1ecb8d7f89eb

leilacc commented on notebooks/bedrock/Financial Multilingual Embeddings.ipynb:
```diff
@@ -0,0 +1,861 @@
+    }
+   ],
+   "source": [
+    "!pip install --upgrade cohere-aws hnswlib translate"
```
Please include a reminder to restart the kernel if any packages are upgraded (you can copy from the other bedrock notebooks)

gonzalocohere replied:
Fixed in 91f7d1cb406b3cadda57438b1e8c1ecb8d7f89eb

leilacc commented on notebooks/bedrock/Financial Multilingual Embeddings.ipynb:
```diff
@@ -0,0 +1,861 @@
+    "\n",
+    "Multfin Github repo is available in https://github.com/RasmusKaer/MultiFin with a link to the data in [Dropbox](https://www.dropbox.com/sh/v0fvtn5gnf4ij6f/AABPSV4NzHIRa8GwUfQ7kA0aa?dl=0)\n",
+    "\n",
+    "Download and unzip the _CPT_Multilingual_Financial.zip_ file. Upload to your Jupyter SageMaker Notebook instance the _train.json_ file available in /MultiFinDataset_EACL/0-all-languages-lowlevel."
```
I found the second part of this instruction confusing. Because you only mention unzipping one file, I assumed that train.json would be in that file. It would be clearer to also mention unzipping MultiFinDataset_EACL

gonzalocohere replied:
I improved the wording, let me know if it's better now. I also realized I was indicating to download the wrong file.

leilacc commented on notebooks/bedrock/Financial Multilingual Embeddings.ipynb:
```diff
@@ -0,0 +1,861 @@
+   "outputs": [],
+   "source": [
+    "# Read the data\n",
+    "df = pd.read_json(\"./train.json\", lines=True)"
```
This is giving me `ValueError: Expected object or value`, any idea why? I have the file.

gonzalocohere replied:
I haven't figured out why. David for example is able to read it without issues. One option (given donwloading the data is a bit cumbersome right now) would be to add the json to our repo and read directly from there. Happy to discuss.

gonzalocohere committed:
Update Financial Multilingual Embeddings.ipynb

Changes:
* Cleared outputs
* Removed trailing I
* Improved explanation to download data
* Included reminder to restart kernel

leilacc commented on notebooks/bedrock/Financial Multilingual Embeddings.ipynb:
```diff
@@ -0,0 +1,545 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Set up translator. You can also commemnt this block and use the Python list at the end of this notebook\n",
```
s/commemnt

gonzalocohere replied:
no longer applicable as I've removed translation block

leilacc commented on notebooks/bedrock/Financial Multilingual Embeddings.ipynb:
```diff
@@ -0,0 +1,545 @@
+   "source": [
+    "# Ensure there is no duplicated text in the headers\n",
+    "def remove_duplicates(text):\n",
+    "    return re.sub(r'((\\b\\w+\\b.{1,2}\\w+\\b)+).+\\1', r'\\1', text, flags=re.I)\n",
```
this is giving me a headache :)
what exactly is going on; are you sure the regex is the best way to accomplish it; and can you add some detail to the comment explaining why there can't be duplicates?

gonzalocohere replied:
Sorry 😬  I'm picking the longest articles, but some are long just because there are repeated sequences over and over. The regex expressions cleans that up. Perhaps I could saw an example first to show more clearly what we're cleaning.

gonzalocohere replied:
Example of cleaning added in commit 67b5f37ca71a4fbfd4a87a9d1671d3319b7b341a

leilacc replied:
thank you!

leilacc commented on notebooks/bedrock/Financial Multilingual Embeddings.ipynb:
```diff
@@ -0,0 +1,545 @@
+    "\n",
+    "# Select languages\n",
+    "languages = ['English', 'Spanish', 'Danish']\n",
+    "top_80_df = df.loc[df['lang'].isin(languages)]\n",
```
nit: these aren't top 80 yet

gonzalocohere replied:
fixed in 67b5f37ca71a4fbfd4a87a9d1671d3319b7b341a

leilacc commented on notebooks/bedrock/Financial Multilingual Embeddings.ipynb:
```diff
@@ -0,0 +1,545 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Set up translator. You can also commemnt this block and use the Python list at the end of this notebook\n",
```
I ran into translation issues and it was a bit of a pain to debug.
What if we just provided the train.json file (to avoid the Dropbox link going down) and included the translated titles in the file to begin with? Then people won't have to worry about this translation stuff.

gonzalocohere replied:
I agree that providing the train.json is better. I don't have translations for the thousands of files though. Let's discuss on Monday.

gonzalocohere replied:
csv file with translations uploaded in commit 3521452eab4a05147372c6449c641e3dc257040d

gonzalocohere replied:
Now the code points to the dataset in the github repo, you just gotta change the branch from 'main' to 'financial-multilingual-embed-notebook' when testing the notebook before merging the request

leilacc replied:
thank you!

leilacc commented on notebooks/bedrock/Financial Multilingual Embeddings.ipynb:
```diff
@@ -0,0 +1,545 @@
+    "doc_embs = co.embed(texts=docs, model_id=model_id, input_type='search_document').embeddings\n",
+    "\n",
+    "# Create a search index\n",
+    "index = hnswlib.Index(space='ip', dim=1024)\n",
```
Can you add comments for some of these magic numbers?

gonzalocohere replied:
more info added in 67b5f37ca71a4fbfd4a87a9d1671d3319b7b341a

leilacc commented on notebooks/bedrock/Financial Multilingual Embeddings.ipynb:
```diff
@@ -0,0 +1,545 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# check language distribution\n",
```
nit: other comments are capitalized

gonzalocohere replied:
fixed in 67b5f37ca71a4fbfd4a87a9d1671d3319b7b341a

leilacc commented on notebooks/bedrock/Financial Multilingual Embeddings.ipynb:
```diff
@@ -0,0 +1,545 @@
+    "languages = ['English', 'Spanish', 'Danish']\n",
+    "top_80_df = df.loc[df['lang'].isin(languages)]\n",
+    "\n",
+    "# Pick the top 80 longest articles\n",
```
Why limit to top 80?

gonzalocohere replied:
I'm trying to stay below text limit per embed call (96 texts) to keep things simple and not have to loop calls.

leilacc replied:
maybe it would be a good teaching moment then? any reason not to use 96 and have a comment saying 'this is the max number of texts per call'?

gonzalocohere replied:
that makes sense - it's just timing, since AWS has been pushing to get this asap and have other stuff I need to get out for reinvent next week 😅

gonzalocohere committed:
Add files via upload

uploading data needed for financial multilingual embed notebook to make process easier for the reader

gonzalocohere committed:
Update Financial Multilingual Embeddings.ipynb

- Simplified data upload and translation management
- Minor formatting fixes
- Improved comments
